halo2_proofs/src/plonk/permutation/keygen.rs

```diff
@@ -108,60 +115,173 @@ impl Assembly {
+    /// Mapping of the actual copies done.
+    aux: HashMap<(usize, usize), usize>,
+    /// total length of a column
+    col_len: usize,
+    /// number of columns
+    num_cols: usize,
+}
+
+#[cfg(feature = "thread-safe-region")]
+impl Assembly {
+    pub(crate) fn new(n: usize, p: &Argument) -> Self {
+        Assembly {
+            columns: p.columns.clone(),
+            cycles: Vec::new(),
+            aux: HashMap::new(),
+            col_len: n,
+            num_cols: p.columns.len(),
         }
+    }
 
-        // Compute [omega_powers * \delta^0, omega_powers * \delta^1, ..., omega_powers * \delta^m]
-        let mut deltaomega = vec![omega_powers; p.columns.len()];
-        {
-            parallelize(&mut deltaomega, |o, start| {
-                let mut cur = C::Scalar::DELTA.pow_vartime(&[start as u64]);
-                for omega_powers in o.iter_mut() {
-                    for v in omega_powers {
-                        *v *= &cur;
-                    }
-                    cur *= &<C::Scalar as PrimeField>::DELTA;
-                }
-            });
+    pub(crate) fn copy(
+        &mut self,
+        left_column: Column<Any>,
+        mut left_row: usize,
+        right_column: Column<Any>,
+        mut right_row: usize,
+    ) -> Result<(), Error> {
+        let mut left_column = self
+            .columns
+            .iter()
+            .position(|c| c == &left_column)
+            .ok_or(Error::ColumnNotInPermutation(left_column))?;
+        let mut right_column = self
+            .columns
+            .iter()
+            .position(|c| c == &right_column)
+            .ok_or(Error::ColumnNotInPermutation(right_column))?;
+
+        // Check bounds
+        if left_row >= self.col_len || right_row >= self.col_len {
+            return Err(Error::BoundsFailure);
         }
 
-        // Computes the permutation polynomial based on the permutation
-        // description in the assembly.
-        let mut permutations = vec![domain.empty_lagrange(); p.columns.len()];
-        {
-            parallelize(&mut permutations, |o, start| {
-                for (x, permutation_poly) in o.iter_mut().enumerate() {
-                    let i = start + x;
-                    for (j, p) in permutation_poly.iter_mut().enumerate() {
-                        let (permuted_i, permuted_j) = self.mapping[i][j];
-                        *p = deltaomega[permuted_i][permuted_j];
-                    }
-                }
-            });
+        // See book/src/design/permutation.md for a description of this algorithm.
+        let mut left_cycle = self.aux.get(&(left_column, left_row));
+        let mut right_cycle = self.aux.get(&(right_column, right_row));
+
+        // If left and right are in the same cycle, do nothing.
+        if (left_cycle == right_cycle) && (left_cycle.is_some()) && (right_cycle.is_some()) {
+            return Ok(());
+        }
+
+        let left_cycle_size = match left_cycle {
```

Comment:
You can use [map_or](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or) here.